### ISSUE TITLE
dynamic batching in effect?

### ISSUE THREAD

**Issue by llxyw** [triton-inference-server/server#6421]
deployed the roformer model and used dynamic batch in config ：
name: "roformer_chinese_sim_char_ft_small"
backend: "tensorrt"
max_batch_size: 64
input {
  name: "input_ids"
  data_type: TYPE_INT32
  dims: [-1]
}
input {
  name: "attention_mask"
  data_type: TYPE_INT32
  dims: [-1]
}
input {
  name: "token_type_ids"
  data_type: TYPE_INT32
  dims: [-1]
}
output {
  name: "pooler_output"
  data_type: TYPE_FP32
  dims: [384]
}

instance_group{
   count:1
   kind:KIND_GPU 
   gpus:[0]
}

dynamic_batching{
   preferred_batch_size:[4,8,16,64]
   max_queue_delay_microseconds:50
}

I wrote a client and sent a request，the tritonserver log：
2023-10-11T05:58:44Z I 7139 pinned_memory_manager.cc:190] pinned memory deallocation: addr 0x7fcb200012c0
2023-10-11T05:58:44Z I 7139 infer_handler.cc:700] Process for ModelInferHandler, rpc_ok=1, 0 step START
2023-10-11T05:58:44Z I 7139 pinned_memory_manager.cc:190] pinned memory deallocation: addr 0x7fcb20000090
2023-10-11T05:58:44Z I 7139 infer_handler.cc:693] New request handler for ModelInferHandler, 0
2023-10-11T05:58:44Z I 7139 pinned_memory_manager.cc:190] pinned memory deallocation: addr 0x7fcb200006a0
2023-10-11T05:58:44Z I 7139 pinned_memory_manager.cc:190] pinned memory deallocation: addr 0x7fcb20000cb0
2023-10-11T05:58:44Z I 7139 infer_request.cc:751] [request id: <id_unknown>] prepared: [0x0x7fcaac006840] request id: , model: roformer_chinese_sim_char_ft_small, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, **batch size: 1**, priority: 0, timeout (us): 0
original inputs:
[0x0x7fcaac00af28] input: token_type_ids, type: INT32, original shape: [1,128], batch + shape: [1,128], shape: [128]
[0x0x7fcaac0060c8] input: input_ids, type: INT32, original shape: [1,128], batch + shape: [1,128], shape: [128]
[0x0x7fcaac00c148] input: attention_mask, type: INT32, original shape: [1,128], batch + shape: [1,128], shape: [128]
override inputs:
inputs:
[0x0x7fcaac00c148] input: attention_mask, type: INT32, original shape: [1,128], batch + shape: [1,128], shape: [128]
[0x0x7fcaac0060c8] input: input_ids, type: INT32, original shape: [1,128], batch + shape: [1,128], shape: [128]
[0x0x7fcaac00af28] input: token_type_ids, type: INT32, original shape: [1,128], batch + shape: [1,128], shape: [128]
original requested outputs:
pooler_output
requested outputs:
pooler_output

Is the batch size printed in this log the requested batch?  Or is it a batch that is fed into the model after waiting for assembly? How do I check whether Dynamic batch takes effect？

**Comment by tanmayv25** [comment]
The batch mentioned here is the batch size as received by server. You can take a look at the `batch_stats` field from the statisitics extension to learn finer details on which batch sizes were executed. 
https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_statistics.md

**Comment by llxyw** [comment]
In other words, dynamic batch processing is not in effect here, and if it was, the batch size printed here would not be 1.

**Comment by llxyw** [comment]
I wrote a client and conducted a pressure test using jmeter, but gpu utilization was not up. Why did dynamic batch not take effect? 

import numpy as np
import tritonclient.http as httpclient
import torch
from roformer import RoFormerForCausalLM, RoFormerConfig
from transformers import BertTokenizer
import uvicorn
from fastapi import BackgroundTasks, FastAPI, Request, Body
from fastapi.responses import JSONResponse, StreamingResponse
from fastapi.exceptions import RequestValidationError

input_name = ['input_ids', 'token_type_ids', 'attention_mask']
output_name = 'pooler_output'
model_name = "roformer_chinese_sim_char_ft_small" 
# pretrained_model = "/root/roformer_chinese_sim_char_ft_small"
pretrained_model = "junnyu/roformer_chinese_sim_char_base"
tokenizer = BertTokenizer.from_pretrained(pretrained_model)

app = FastAPI(docs_url='/')

# 自定义异常处理
@app.exception_handler(RequestValidationError)  # 重写了RequestValidationError的 exception_handler方法
async def post_validation_exception_handler(request: Request, exc: RequestValidationError):

    """
    自定义异常处理
    :param request: Request的实例化对象
    :param exc: RequestValidationError的错误堆栈信息
    :return:
    """

    print(f'参数不对{request.method},{request.url}')
    return JSONResponse({'code': 400, 'msg': exc.errors()})

@app.post('/generate')
async def generate(text:str=Body(None,title='text',max_length=1000)):
    triton_client = httpclient.InferenceServerClient(url='0.0.0.0:9003')
    
    tokens = tokenizer(text, return_tensors="pt",max_length=128, truncation=True, padding=True)

    input_ids=tokens["input_ids"]
    input_ids = np.array(input_ids, dtype=np.int32)
    token_type_ids=tokens["token_type_ids"]
    token_type_ids = np.array(token_type_ids, dtype=np.int32)
    attention_mask=tokens["attention_mask"]
    attention_mask = np.array(attention_mask, dtype=np.int32)
    
    input_ids_shape = input_ids.shape
    attention_mask_shape = attention_mask.shape
    token_type_ids_shape = token_type_ids.shape
    
    input0 = httpclient.InferInput(input_name[0], input_ids_shape, 'INT32')
    input0.set_data_from_numpy(input_ids, binary_data=False)
    input1 = httpclient.InferInput(input_name[1], token_type_ids_shape, 'INT32')
    input1.set_data_from_numpy(token_type_ids, binary_data=False)
    input2 = httpclient.InferInput(input_name[2], attention_mask_shape, 'INT32')
    input2.set_data_from_numpy(attention_mask, binary_data=False)
    
    output = httpclient.InferRequestedOutput(output_name,binary_data=False)
    response = triton_client.infer(model_name,inputs=[input0,input1,input2], outputs=[output])
    ans_numpy = response.as_numpy('pooler_output')
    result = ans_numpy.tolist()
    if not result:
         abort(400, description="Inference result is empty")

    return result
    
if __name__ == '__main__':
    uvicorn.run(app=app, host='0.0.0.0', port=9110, log_level='info')

**Comment by Gcstk** [comment]
I have the same issue as described above. I have deployed a text2vec model and tested its performance using the perf analyzer tool. The GPU utilization can reach up to 90% with a single instance. However, when I use Python multi-threading to call the HTTP client API with the same configuration, the GPU utilization drops to just 3%, and the throughput is also very low.

I would like to know where the issue could possibly be:
Is it that Python's multi-threading is not as efficient as the multi-threading in C++ used by perf?
Or is it that dynamic batching is not properly set up?
Is there a log where I can see the actual batch_size that is being fed into the model?
Thank you for your assistance!

**Comment by llxyw** [comment]
How can I see if dynamic batch processing is working

**Comment by dyastremsky** [comment]
Closing this issue, as Tanmay has answered the original question.

Geotk, please open a new issue following the bug template to allow us to see and reproduce all your steps. Your issue does not seem related. We can properly assist you once we see and reproduce what you are doing.
